Update to support Python 3.11 and 3.12

Rana-KV opened the pull request:
## Description

**Addresses Issue**: #379
fixes: #379
**Summary of Changes**:
1. **Dependency Updates**:
   - Updated dependency versions to support Python 3.11 and Python 3.12.
   - The build process will now consider the Python version to select the appropriate libraries.
   - **Click Library**: Updated to version 8.*.*.
   - **pygit2 Library**:
     - For Python 3.8, updated to version 1.10.1.
     - For Python 3.9 and 3.10, updated to version 1.13.3.
     - For Python 3.11 and 3.12, updated to version 1.14.1.

2. **Build and Workflow Updates**:
   - Updated GitHub workflows to accommodate the new Python versions and dependencies.
   - Updated Azure Pipelines configuration for compatibility with the updated dependencies and Python versions.

3. **Documentation**:
   - Updated `CHANGELOG.md` to reflect these changes.

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/taf/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/

renatav reviewed (commented):
Looks like flake 8 checks are more thorough for Python 3.12, so the build failed. See https://github.com/openlawlibrary/taf/actions/runs/9407771803/job/25914281561?pr=440
Could you take a look at those issues. They seem trivial.

We also always update this [CHANGELOG ](https://github.com/openlawlibrary/taf/blob/master/CHANGELOG.md) file before merging a PR which introducing changes worth noting.

Rana-KV commented:
Hi @renatav, I have fixed few of the errors and added trivial errors (like space after ':' ) in ignore list.
I have updated the [CHANGELOG.md](https://github.com/openlawlibrary/taf/blob/master/CHANGELOG.md) in the commits mentioning that Python 3.11 and 3.12 are supported. Should I keep them or remove it?

renatav commented on setup.py:
```diff
@@ -46,6 +47,14 @@ def finalize_options(self):
 yubikey_require = ["yubikey-manager==4.0.*"]
 
+# Determine the appropriate version of pygit2 based on the Python version
+if sys.version_info >= (3, 9):
```
Had to check something, but there was a reason why `pygit2` was pinned to `1.9`. A git security check was added in 1.9.1,  which broke something in our system. It'll need to stay pinned. Since our system is still on Python 3.10, we can set different versions of `pygit2` in case of 3.11 and 3.12 if that is necessary

Rana-KV replied:
Sure, I can pin the version for `Python 3.10`. But the testcases used in pytest didn't raise any error when I ran the TAF builds with latest versions of `pygit2` for `Python 3.10`. On the other side `Python 3.11` needs `pygit2 version >=1.11.x` and `Python 3.12` needs `pygit2 version >=1.13.x` 

Should I just change it for  `Python 3.10` ?

renatav replied:
Let's pin it for `Python <= 3.10`. The error is not happening in TAF, it's in a component of our system which uses TAF. We'll update it if once we've addressed the core issue.

renatav commented:
@Rana-KV We definitely want to add this to the changelog. Could you please also link a PR there? Take a look at other changelog entries. We sort PRs in the descending order. Thanks for looking into the flake issues!
